File: runtimes/http-ts/test/execute.test.ts
import { describe, expect, it } from "vitest";

import { execute } from "../src/execute";
import type { Part, WireDescriptor } from "../src/descriptor";

// A recording transport: captures the one request it is given and returns a
// canned response, so a test can assert exactly what the runtime built.
function recorder(status = 200, body = "{}", headers: Record<string, string> = {}) {
  const calls: { url: string; init: RequestInit }[] = [];
  const fetchImpl = (async (url: string | URL | Request, init?: RequestInit) => {
    calls.push({ url: String(url), init: init ?? {} });
    return new Response(body, { status, headers });
  }) as unknown as typeof fetch;
  return { calls, fetchImpl };
}

function descriptor(over: Partial<WireDescriptor> = {}): WireDescriptor {
  return {
    http_method: "POST",
    uri: "/things",
    bindings: [],
    response_bindings: [],
    success: [[200, null]],
    ...over,
  };
}

const binding = (name: string, part: Part): readonly [string, Part] => [name, part];

describe("request building across the five part variants", () => {
  it("substitutes a label into the uri path", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ http_method: "GET", uri: "/things/{id}", bindings: [binding("id", { kind: "label" })] }),
      { id: "abc" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/things/abc");
  });

  it("substitutes an absent label as empty, never the literal undefined", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ http_method: "GET", uri: "/things/{id}", bindings: [binding("id", { kind: "label" })] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/things/");
  });

  it("appends query members, repeating a list", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({
        http_method: "GET",
        uri: "/things",
        bindings: [binding("q", { kind: "query", name: "q" }), binding("tag", { kind: "query", name: "tag" })],
      }),
      { q: "hi", tag: ["a", "b"] },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/things?q=hi&tag=a&tag=b");
  });

  it("omits a null query member", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ http_method: "GET", uri: "/things", bindings: [binding("q", { kind: "query", name: "q" })] }),
      { q: null },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/things");
  });

  it("sets a header member", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ bindings: [binding("trace", { kind: "header", name: "X-Trace" })] }),
      { trace: "t1" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect((calls[0].init.headers as Record<string, string>)["X-Trace"]).toBe("t1");
  });

  it("assembles body members into a JSON object", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ bindings: [binding("a", { kind: "body" }), binding("b", { kind: "body" })] }),
      { a: 1, b: "two" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].init.body).toBe(JSON.stringify({ a: 1, b: "two" }));
    expect((calls[0].init.headers as Record<string, string>)["content-type"]).toBe("application/json");
  });

  it("sends the single payload member as the whole body, no envelope", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ bindings: [binding("raw", { kind: "payload" })] }),
      { raw: { nested: true } },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].init.body).toBe(JSON.stringify({ nested: true }));
  });

  it("does not add a second content-type when the caller set one under a different case", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ bindings: [binding("a", { kind: "body" })] }),
      { a: 1 },
      { baseUrl: "https://api.test", fetch: fetchImpl, headers: { "Content-Type": "application/vnd.api+json" } },
    );
    const headers = calls[0].init.headers as Record<string, string>;
    expect(headers["Content-Type"]).toBe("application/vnd.api+json");
    expect(headers["content-type"]).toBeUndefined();
  });
});

describe("transport call", () => {
  it("issues the call with the descriptor's method and built uri", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ http_method: "DELETE", uri: "/things/{id}", bindings: [binding("id", { kind: "label" })] }),
      { id: "42" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].init.method).toBe("DELETE");
    expect(calls[0].url).toBe("https://api.test/things/42");
  });
});

describe("response classification", () => {
  it("maps a declared success status to a success outcome", async () => {
    const { fetchImpl } = recorder(200, '{"id":"x"}');
    const outcome = await execute(descriptor(), {}, { baseUrl: "https://api.test", fetch: fetchImpl });
    expect(outcome).toEqual({ outcome: "success", status: 200, body: '{"id":"x"}' });
  });

  it("treats any 2xx as success even when its code was not the declared one", async () => {
    // The descriptor declares 201, the server answers 200: still a success.
    const { fetchImpl } = recorder(200, '{"ok":true}');
    const outcome = await execute(
      descriptor({ success: [[201, null]] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(outcome).toEqual({ outcome: "success", status: 200, body: '{"ok":true}' });
  });

  it("maps a non-success status to an error outcome for the SDK to discriminate", async () => {
    const { fetchImpl } = recorder(404, '{"code":"not_found"}');
    const outcome = await execute(descriptor(), {}, { baseUrl: "https://api.test", fetch: fetchImpl });
    expect(outcome).toEqual({ outcome: "error", status: 404, body: '{"code":"not_found"}' });
  });

  it("reports a network failure as a transport outcome", async () => {
    const boom = new Error("network down");
    const fetchImpl = (async () => {
      throw boom;
    }) as unknown as typeof fetch;
    const outcome = await execute(descriptor(), {}, { baseUrl: "https://api.test", fetch: fetchImpl });
    expect(outcome).toEqual({ outcome: "transport", cause: boom });
  });

  it("reports a body-read failure as a transport outcome", async () => {
    const boom = new Error("stream aborted");
    const fetchImpl = (async () =>
      ({
        status: 200,
        text: async () => {
          throw boom;
        },
        headers: new Headers(),
      }) as unknown as Response) as unknown as typeof fetch;
    const outcome = await execute(descriptor(), {}, { baseUrl: "https://api.test", fetch: fetchImpl });
    expect(outcome).toEqual({ outcome: "transport", cause: boom });
  });
});

describe("response bindings", () => {
  it("drops a non-JSON body and lets the bound fields stand on their own", async () => {
    const { fetchImpl } = recorder(200, "not json", { "X-Request-Id": "req-1" });
    const outcome = await execute(
      descriptor({ response_bindings: [["requestId", { kind: "header", name: "X-Request-Id" }]] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(outcome).toEqual({
      outcome: "success",
      status: 200,
      body: JSON.stringify({ requestId: "req-1" }),
    });
  });

  it("folds a response header into the decoded body under its member name", async () => {
    const { fetchImpl } = recorder(200, '{"id":"x"}', { "X-Request-Id": "req-1" });
    const outcome = await execute(
      descriptor({ response_bindings: [["requestId", { kind: "header", name: "X-Request-Id" }]] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(outcome).toEqual({
      outcome: "success",
      status: 200,
      body: JSON.stringify({ id: "x", requestId: "req-1" }),
    });
  });

  it("folds the response status code into the decoded body", async () => {
    const { fetchImpl } = recorder(200, '{"id":"x"}');
    const outcome = await execute(
      descriptor({ response_bindings: [["httpStatus", { kind: "statusCode" }]] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(outcome).toEqual({
      outcome: "success",
      status: 200,
      body: JSON.stringify({ id: "x", httpStatus: 200 }),
    });
  });
});

describe("verbatim execution", () => {
  it("applies no binding defaults of its own: an explicit descriptor produces exactly that request", async () => {
    const { calls, fetchImpl } = recorder();
    // Every member is explicitly bound; the runtime must not reclassify any.
    await execute(
      descriptor({
        http_method: "PUT",
        uri: "/x/{id}",
        bindings: [
          binding("id", { kind: "label" }),
          binding("q", { kind: "query", name: "q" }),
          binding("h", { kind: "header", name: "H" }),
          binding("field", { kind: "body" }),
        ],
      }),
      { id: "1", q: "2", h: "3", field: "4" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/x/1?q=2");
    expect((calls[0].init.headers as Record<string, string>)["H"]).toBe("3");
    expect(calls[0].init.body).toBe(JSON.stringify({ field: "4" }));
  });
});

describe("edge cases pinned by the mutation gate", () => {
  it("normalizes a null input to an empty record instead of dereferencing it", async () => {
    // A null input with a body binding must not be indexed into; it yields no body.
    const { calls, fetchImpl } = recorder();
    const outcome = await execute(
      descriptor({ bindings: [binding("a", { kind: "body" })] }),
      null,
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(outcome.outcome).toBe("success");
    expect(calls[0].init.body).toBeUndefined();
  });

  it("normalizes a non-object input to an empty record", async () => {
    // A string input has no members: a query binding named after a string
    // property ("length") must not leak the string's own length.
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({
        http_method: "GET",
        uri: "/things",
        bindings: [binding("length", { kind: "query", name: "length" })],
      }),
      "hello",
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/things");
  });

  it("leaves a non-label binding out of path substitution", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({
        http_method: "GET",
        uri: "/things/{id}",
        bindings: [binding("id", { kind: "label" }), binding("q", { kind: "query", name: "q" })],
      }),
      { id: "7", q: "x" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/things/7?q=x");
  });

  it("substitutes a present label rather than emptying it", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ http_method: "GET", uri: "/things/{id}", bindings: [binding("id", { kind: "label" })] }),
      { id: "present" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/things/present");
  });

  it("omits an absent query member and keeps a present one", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({
        http_method: "GET",
        uri: "/things",
        bindings: [binding("q", { kind: "query", name: "q" }), binding("r", { kind: "query", name: "r" })],
      }),
      { q: "kept" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/things?q=kept");
  });

  it("omits a null header and keeps a present one", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({
        bindings: [
          binding("keep", { kind: "header", name: "X-Keep" }),
          binding("drop", { kind: "header", name: "X-Drop" }),
        ],
      }),
      { keep: "yes", drop: null },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    const headers = calls[0].init.headers as Record<string, string>;
    expect(headers["X-Keep"]).toBe("yes");
    expect(headers["X-Drop"]).toBeUndefined();
  });

  it("drops an absent body member from the assembled object", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ bindings: [binding("a", { kind: "body" }), binding("b", { kind: "body" })] }),
      { a: 1 },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].init.body).toBe(JSON.stringify({ a: 1 }));
  });

  it("sends no body and no content-type when every body member is absent", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ bindings: [binding("a", { kind: "body" })] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].init.body).toBeUndefined();
    expect((calls[0].init.headers as Record<string, string>)["content-type"]).toBeUndefined();
  });

  it("adds the default content-type when a body is present and none was set", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ bindings: [binding("a", { kind: "body" })] }),
      { a: 1 },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect((calls[0].init.headers as Record<string, string>)["content-type"]).toBe("application/json");
  });

  it("suppresses the default content-type when the caller set it in exact lower case", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ bindings: [binding("a", { kind: "body" })] }),
      { a: 1 },
      { baseUrl: "https://api.test", fetch: fetchImpl, headers: { "content-type": "text/plain" } },
    );
    expect((calls[0].init.headers as Record<string, string>)["content-type"]).toBe("text/plain");
  });

  it("classifies the 2xx boundary: 299 succeeds, 300 errors", async () => {
    const ok = recorder(299, "{}");
    expect(
      (await execute(descriptor(), {}, { baseUrl: "https://api.test", fetch: ok.fetchImpl })).outcome,
    ).toBe("success");
    const bad = recorder(300, "{}");
    expect(
      (await execute(descriptor(), {}, { baseUrl: "https://api.test", fetch: bad.fetchImpl })).outcome,
    ).toBe("error");
  });

  it("honors a declared success code outside the 2xx range", async () => {
    const { fetchImpl } = recorder(302, "{}");
    const outcome = await execute(
      descriptor({ success: [[302, null]] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(outcome.outcome).toBe("success");
  });

  it("does not treat an undeclared non-2xx code as success", async () => {
    const { fetchImpl } = recorder(418, "{}");
    const outcome = await execute(
      descriptor({ success: [[302, null]] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(outcome.outcome).toBe("error");
  });

  it("returns the body verbatim when there are no response bindings", async () => {
    // A spaced body proves the text is passed through, not re-serialized: the
    // early return must fire when there is nothing to fold in.
    const { fetchImpl } = recorder(200, '{"id": "x"}');
    const outcome = await execute(descriptor(), {}, { baseUrl: "https://api.test", fetch: fetchImpl });
    expect(outcome).toEqual({ outcome: "success", status: 200, body: '{"id": "x"}' });
  });

  it("only substitutes label bindings into the path, not query bindings", async () => {
    // A query binding whose name matches a path placeholder must be left as a
    // literal: only labels feed path substitution.
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ http_method: "GET", uri: "/things/{q}", bindings: [binding("q", { kind: "query", name: "q" })] }),
      { q: "v" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/things/{q}?q=v");
  });

  it("substitutes a null label as empty rather than the literal null", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ http_method: "GET", uri: "/things/{id}", bindings: [binding("id", { kind: "label" })] }),
      { id: null },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(calls[0].url).toBe("https://api.test/things/");
  });

  it("only reads header bindings into headers, not query bindings", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ http_method: "GET", uri: "/things", bindings: [binding("q", { kind: "query", name: "q" })] }),
      { q: "v" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect((calls[0].init.headers as Record<string, string>)["q"]).toBeUndefined();
  });

  it("omits an absent header member rather than sending the literal undefined", async () => {
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ bindings: [binding("trace", { kind: "header", name: "X-Trace" })] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect((calls[0].init.headers as Record<string, string>)["X-Trace"]).toBeUndefined();
  });

  it("adds the default content-type even when another header is already present", async () => {
    // hasHeader must match the content-type name specifically, not just report
    // that some header exists.
    const { calls, fetchImpl } = recorder();
    await execute(
      descriptor({ bindings: [binding("a", { kind: "body" }), binding("trace", { kind: "header", name: "X-Trace" })] }),
      { a: 1, trace: "t" },
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    const headers = calls[0].init.headers as Record<string, string>;
    expect(headers["X-Trace"]).toBe("t");
    expect(headers["content-type"]).toBe("application/json");
  });

  it("requires a 2xx floor: a sub-200 status is not a success on its own", async () => {
    // The Response constructor forbids a sub-200 status, so a bare stub stands in
    // to exercise the lower bound of the 2xx check.
    const fetchImpl = (async () =>
      ({ status: 150, text: async () => "{}", headers: new Headers() }) as unknown as Response) as unknown as typeof fetch;
    const outcome = await execute(descriptor(), {}, { baseUrl: "https://api.test", fetch: fetchImpl });
    expect(outcome.outcome).toBe("error");
  });

  it("accepts a status matching any one declared success code, not all of them", async () => {
    const { fetchImpl } = recorder(302, "{}");
    const outcome = await execute(
      descriptor({ success: [[301, null], [302, null]] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(outcome.outcome).toBe("success");
  });

  it("stands a response-bound field on its own when the body is empty", async () => {
    const { fetchImpl } = recorder(200, "", { "X-Id": "r1" });
    const outcome = await execute(
      descriptor({ response_bindings: [["id", { kind: "header", name: "X-Id" }]] }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(outcome).toEqual({ outcome: "success", status: 200, body: JSON.stringify({ id: "r1" }) });
  });
});

File: runtimes/http-go/execute_test.go
package tonohttp

import (
	"context"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func newRuntime(t *testing.T, opts Options) *Runtime {
	t.Helper()
	r, err := New(opts)
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	return r
}

// A recording transport: captures every request and answers from a fixed
// response, so a test can assert exactly what the runtime built.
func recorder(status int, body string, headers map[string]string) (Transport, *[]CanonicalRequest) {
	var calls []CanonicalRequest
	transport := func(ctx context.Context, req CanonicalRequest) (CanonicalResponse, error) {
		calls = append(calls, req)
		return CanonicalResponse{Status: status, Headers: headers, Body: body}, nil
	}
	return transport, &calls
}

func TestNewRejectsBothTransportSlots(t *testing.T) {
	transport, _ := recorder(200, "{}", nil)
	if _, err := New(Options{Client: &http.Client{}, Transport: transport}); err == nil {
		t.Fatal("both slots set must be a construction error")
	} else if !strings.Contains(err.Error(), "mutually exclusive") {
		t.Fatalf("error must name the conflict: %v", err)
	}
	if _, err := New(Options{Client: &http.Client{}}); err != nil {
		t.Fatalf("native slot alone: %v", err)
	}
	if _, err := New(Options{Transport: transport}); err != nil {
		t.Fatalf("canonical slot alone: %v", err)
	}
	if _, err := New(Options{}); err != nil {
		t.Fatalf("neither slot (default client): %v", err)
	}
}

func TestCanonicalTransportReceivesTheBuiltRequest(t *testing.T) {
	transport, calls := recorder(200, "{}", nil)
	r := newRuntime(t, Options{
		BaseURL:   "https://api.test",
		Transport: transport,
		Headers:   map[string]string{"Authorization": "Bearer t"},
	})
	d := desc(func(d *WireDescriptor) {
		d.HTTPMethod = "PUT"
		d.URI = "/x/{id}"
		d.Bindings = []Binding{
			binding("id", "label", ""),
			binding("q", "query", "q"),
			binding("h", "header", "H"),
			binding("field", "body", ""),
		}
	})
	outcome, err := r.Execute(context.Background(), d, map[string]any{"id": "1", "q": "2", "h": "3", "field": "4"}, nil)
	if err != nil || outcome.Kind != OutcomeSuccess {
		t.Fatalf("outcome: %+v %v", outcome, err)
	}
	if len(*calls) != 1 {
		t.Fatalf("attempts: %d", len(*calls))
	}
	req := (*calls)[0]
	if req.Method != "PUT" || req.URL != "https://api.test/x/1?q=2" {
		t.Fatalf("request line: %s %s", req.Method, req.URL)
	}
	if req.Headers["H"] != "3" || req.Headers["Authorization"] != "Bearer t" {
		t.Fatalf("headers: %+v", req.Headers)
	}
	if string(req.Body) != `{"field":"4"}` {
		t.Fatalf("body: %s", req.Body)
	}
	if req.Headers["content-type"] != "application/json" {
		t.Fatalf("default content-type missing: %+v", req.Headers)
	}
}

func TestContentTypeDefault(t *testing.T) {
	d := desc(func(d *WireDescriptor) {
		d.Bindings = []Binding{binding("a", "body", "")}
	})

	transport, calls := recorder(200, "{}", nil)
	r := newRuntime(t, Options{BaseURL: "https://api.test", Transport: transport})
	if _, err := r.Execute(context.Background(), d, map[string]any{}, nil); err != nil {
		t.Fatal(err)
	}
	if _, ok := (*calls)[0].Headers["content-type"]; ok || (*calls)[0].Body != nil {
		t.Fatalf("no body must mean no content-type: %+v", (*calls)[0])
	}

	transport, calls = recorder(200, "{}", nil)
	r = newRuntime(t, Options{
		BaseURL:   "https://api.test",
		Transport: transport,
		Headers:   map[string]string{"Content-Type": "application/vnd.api+json"},
	})
	if _, err := r.Execute(context.Background(), d, map[string]any{"a": float64(1)}, nil); err != nil {
		t.Fatal(err)
	}
	headers := (*calls)[0].Headers
	if headers["Content-Type"] != "application/vnd.api+json" {
		t.Fatalf("caller content-type lost: %+v", headers)
	}
	if _, ok := headers["content-type"]; ok {
		t.Fatalf("second content-type added beside the caller's: %+v", headers)
	}
}

func TestNativeSlotAgainstARealServer(t *testing.T) {
	var seen *http.Request
	var seenBody string
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		seen = req.Clone(context.Background())
		data, _ := io.ReadAll(req.Body)
		seenBody = string(data)
		w.Header().Set("X-Request-Id", "req-1")
		w.WriteHeader(200)
		_, _ = w.Write([]byte(`{"id":"x"}`))
	}))
	defer server.Close()

	// Client nil exercises the http.DefaultClient fallback.
	r := newRuntime(t, Options{BaseURL: server.URL})
	d := desc(func(d *WireDescriptor) {
		d.URI = "/things/{id}"
		d.Bindings = []Binding{
			binding("id", "label", ""),
			binding("q", "query", "q"),
			binding("trace", "header", "X-Trace"),
			binding("a", "body", ""),
		}
		d.ResponseBindings = []ResponseBinding{{Member: "requestId", Part: ResponsePart{Kind: "header", Name: "X-Request-Id"}}}
	})
	outcome, err := r.Execute(context.Background(), d, map[string]any{"id": "7", "q": "v", "trace": "t1", "a": float64(1)}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if outcome.Kind != OutcomeSuccess || outcome.Status != 200 {
		t.Fatalf("outcome: %+v", outcome)
	}
	// The response-bound header folds into the body under its member name.
	if outcome.Body != `{"id":"x","requestId":"req-1"}` {
		t.Fatalf("folded body: %s", outcome.Body)
	}
	if seen.Method != "POST" || seen.URL.Path != "/things/7" || seen.URL.RawQuery != "q=v" {
		t.Fatalf("request line: %s %s?%s", seen.Method, seen.URL.Path, seen.URL.RawQuery)
	}
	if seen.Header.Get("X-Trace") != "t1" || seen.Header.Get("Content-Type") != "application/json" {
		t.Fatalf("headers: %+v", seen.Header)
	}
	if seenBody != `{"a":1}` {
		t.Fatalf("body: %s", seenBody)
	}
}

func TestNativeTransportFailures(t *testing.T) {
	r := newRuntime(t, Options{BaseURL: "http://127.0.0.1:0"})
	outcome, err := r.Execute(context.Background(), desc(nil), nil, nil)
	if err != nil || outcome.Kind != OutcomeTransport || outcome.Cause == nil {
		t.Fatalf("network failure must be a transport outcome: %+v %v", outcome, err)
	}

	// An invalid method fails request construction, which is transport too.
	d := desc(func(d *WireDescriptor) { d.HTTPMethod = "BAD METHOD" })
	outcome, err = r.Execute(context.Background(), d, nil, nil)
	if err != nil || outcome.Kind != OutcomeTransport {
		t.Fatalf("bad method: %+v %v", outcome, err)
	}
}

func TestNativeBodyReadFailureIsTransport(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		// Promise more bytes than are sent, then cut the connection: the body
		// read fails mid-stream.
		w.Header().Set("Content-Length", "100")
		_, _ = w.Write([]byte("short"))
		if f, ok := w.(http.Flusher); ok {
			f.Flush()
		}
		hj, ok := w.(http.Hijacker)
		if !ok {
			return
		}
		conn, _, _ := hj.Hijack()
		conn.Close()
	}))
	defer server.Close()
	r := newRuntime(t, Options{BaseURL: server.URL, Client: server.Client()})
	outcome, err := r.Execute(context.Background(), desc(nil), nil, nil)
	if err != nil || outcome.Kind != OutcomeTransport || outcome.Cause == nil {
		t.Fatalf("body-read failure must be a transport outcome: %+v %v", outcome, err)
	}
}

func TestSuccessClassification(t *testing.T) {
	run := func(status int, mutate func(*WireDescriptor)) Outcome {
		transport, _ := recorder(status, "{}", nil)
		r := newRuntime(t, Options{BaseURL: "https://api.test", Transport: transport})
		outcome, err := r.Execute(context.Background(), desc(mutate), nil, nil)
		if err != nil {
			t.Fatal(err)
		}
		return outcome
	}
	if run(299, nil).Kind != OutcomeSuccess {
		t.Fatal("299 is a 2xx success")
	}
	if run(300, nil).Kind != OutcomeError {
		t.Fatal("300 is not a success")
	}
	if run(150, nil).Kind != OutcomeError {
		t.Fatal("a sub-200 status is not a success on its own")
	}
	if run(200, func(d *WireDescriptor) { d.Success = []SuccessCase{{Status: 201}} }).Kind != OutcomeSuccess {
		t.Fatal("any 2xx succeeds even when another code was declared")
	}
	declared := func(d *WireDescriptor) { d.Success = []SuccessCase{{Status: 301}, {Status: 302}} }
	if run(302, declared).Kind != OutcomeSuccess {
		t.Fatal("a declared non-2xx code succeeds")
	}
	if run(418, declared).Kind != OutcomeError {
		t.Fatal("an undeclared non-2xx code errors")
	}
}

func TestResponseBindings(t *testing.T) {
	run := func(body string, headers map[string]string, mutate func(*WireDescriptor)) Outcome {
		transport, _ := recorder(200, body, headers)
		r := newRuntime(t, Options{BaseURL: "https://api.test", Transport: transport})
		outcome, err := r.Execute(context.Background(), desc(mutate), nil, nil)
		if err != nil {
			t.Fatal(err)
		}
		return outcome
	}
	headerBinding := func(d *WireDescriptor) {
		d.ResponseBindings = []ResponseBinding{{Member: "requestId", Part: ResponsePart{Kind: "header", Name: "X-Request-Id"}}}
	}
	// Canonical response headers are lowercased by contract; the lookup is too.
	got := run(`{"id":"x"}`, map[string]string{"x-request-id": "r1"}, headerBinding)
	if got.Body != `{"id":"x","requestId":"r1"}` {
		t.Fatalf("header fold: %s", got.Body)
	}
	got = run("", map[string]string{"x-request-id": "r1"}, headerBinding)
	if got.Body != `{"requestId":"r1"}` {
		t.Fatalf("bound field must stand alone on an empty body: %s", got.Body)
	}
	got = run("not json", nil, headerBinding)
	if got.Body != `{"requestId":null}` {
		t.Fatalf("missing header folds as null, non-JSON body dropped: %s", got.Body)
	}
	got = run(`{"id":"x"}`, nil, func(d *WireDescriptor) {
		d.ResponseBindings = []ResponseBinding{{Member: "httpStatus", Part: ResponsePart{Kind: "statusCode"}}}
	})
	if got.Body != `{"httpStatus":200,"id":"x"}` {
		t.Fatalf("status fold: %s", got.Body)
	}
	// A spaced body proves the text passes through verbatim when there is
	// nothing to fold in.
	got = run(`{"id": "x"}`, nil, nil)
	if got.Body != `{"id": "x"}` {
		t.Fatalf("verbatim body: %s", got.Body)
	}
}

func TestHooks(t *testing.T) {
	transport, calls := recorder(200, "{}", nil)
	r := newRuntime(t, Options{BaseURL: "https://api.test", Transport: transport})
	hooks := &Hooks{
		BeforeRequest: func(ctx context.Context, req CanonicalRequest) (CanonicalRequest, error) {
			req.Headers["Authorization"] = "signed"
			return req, nil
		},
	}
	if _, err := r.Execute(context.Background(), desc(nil), nil, hooks); err != nil {
		t.Fatal(err)
	}
	if (*calls)[0].Headers["Authorization"] != "signed" {
		t.Fatalf("BeforeRequest rewrite lost: %+v", (*calls)[0].Headers)
	}

	transport, _ = recorder(500, "{}", nil)
	r = newRuntime(t, Options{BaseURL: "https://api.test", Transport: transport})
	rewrite := &Hooks{
		AfterResponse: func(ctx context.Context, res CanonicalResponse) (CanonicalResponse, error) {
			res.Status = 200
			return res, nil
		},
	}
	outcome, err := r.Execute(context.Background(), desc(nil), nil, rewrite)
	if err != nil || outcome.Kind != OutcomeSuccess {
		t.Fatalf("classification must read the post-hook response: %+v %v", outcome, err)
	}
}

func TestHookErrorsPropagateRawAndNeverRetry(t *testing.T) {
	boom := errors.New("hook broke")
	withRetry := func(d *WireDescriptor) { d.Retry = &RetrySpec{Max: ValueSource{Lit: lit(3)}} }

	transport, calls := recorder(200, "{}", nil)
	r := newRuntime(t, Options{BaseURL: "https://api.test", Transport: transport})
	_, err := r.Execute(context.Background(), desc(withRetry), nil, &Hooks{
		BeforeRequest: func(ctx context.Context, req CanonicalRequest) (CanonicalRequest, error) {
			return req, boom
		},
	})
	if !errors.Is(err, boom) || len(*calls) != 0 {
		t.Fatalf("BeforeRequest error must propagate before any attempt: %v, %d calls", err, len(*calls))
	}

	transport, calls = recorder(503, "{}", nil)
	r = newRuntime(t, Options{BaseURL: "https://api.test", Transport: transport})
	_, err = r.Execute(context.Background(), desc(func(d *WireDescriptor) {
		withRetry(d)
		d.Errors = []DeclaredError{{Status: 503, ID: "svc#unavailable", Retryable: true}}
	}), nil, &Hooks{
		AfterResponse: func(ctx context.Context, res CanonicalResponse) (CanonicalResponse, error) {
			return res, boom
		},
	})
	if !errors.Is(err, boom) || len(*calls) != 1 {
		t.Fatalf("AfterResponse error must propagate, not retry: %v, %d calls", err, len(*calls))
	}
}

func TestUnencodableInputSurfacesAsAnError(t *testing.T) {
	transport, calls := recorder(200, "{}", nil)
	r := newRuntime(t, Options{BaseURL: "https://api.test", Transport: transport})
	d := desc(func(d *WireDescriptor) {
		d.Bindings = []Binding{binding("a", "body", "")}
	})
	if _, err := r.Execute(context.Background(), d, map[string]any{"a": make(chan int)}, nil); err == nil {
		t.Fatal("unencodable input must error, not call the transport")
	}
	if len(*calls) != 0 {
		t.Fatalf("transport called %d times", len(*calls))
	}
}

func TestTimeoutDeadlineIsSetOnlyWhenDeclared(t *testing.T) {
	var hadDeadline bool
	transport := func(ctx context.Context, req CanonicalRequest) (CanonicalResponse, error) {
		_, hadDeadline = ctx.Deadline()
		return CanonicalResponse{Status: 200, Body: "{}"}, nil
	}
	r := newRuntime(t, Options{BaseURL: "https://api.test", Transport: transport})
	if _, err := r.Execute(context.Background(), desc(nil), nil, nil); err != nil {
		t.Fatal(err)
	}
	if hadDeadline {
		t.Fatal("no declared timeout must mean no deadline")
	}
	d := desc(func(d *WireDescriptor) { d.Timeout = &ValueSource{Lit: lit(5000)} })
	if _, err := r.Execute(context.Background(), d, nil, nil); err != nil {
		t.Fatal(err)
	}
	if !hadDeadline {
		t.Fatal("declared timeout must set a deadline")
	}
}
